File: test/mbc_dispatch_catalog_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:psyclinicai/services/assessments/outcome_measure_catalog.dart';
import 'package:psyclinicai/services/mbc/mbc_dispatch_catalog.dart';

void main() {
  group('MbcDispatchCatalog invariants', () {
    test('catalog is non-empty + every rule has a known scale id', () {
      expect(MbcDispatchCatalog.rules, isNotEmpty);
      for (final r in MbcDispatchCatalog.rules) {
        final outcome = OutcomeMeasureCatalog.byScaleId(r.scaleId);
        expect(outcome, isNotNull,
            reason: '${r.scaleId}: unknown to OutcomeMeasureCatalog');
      }
    });

    test('intervalDays matches OutcomeMeasureCatalog.readminInterval', () {
      for (final r in MbcDispatchCatalog.rules) {
        final outcome = OutcomeMeasureCatalog.byScaleId(r.scaleId)!;
        expect(r.intervalDays, outcome.readminInterval,
            reason: '${r.scaleId}: drift vs OutcomeMeasureCatalog');
      }
    });

    test('reminderAtHours <= linkLifetimeHours (reminders never dead)', () {
      for (final r in MbcDispatchCatalog.rules) {
        expect(r.reminderAtHours <= r.linkLifetimeHours, isTrue,
            reason: '${r.scaleId}: reminder would point at expired link');
      }
    });

    test('every rule pins at least one audience + channel', () {
      for (final r in MbcDispatchCatalog.rules) {
        expect(r.audiences, isNotEmpty);
        expect(r.channels, isNotEmpty);
      }
    });

    test('every rule pins a non-empty payerCadenceLabel', () {
      for (final r in MbcDispatchCatalog.rules) {
        expect(r.payerCadenceLabel.trim(), isNotEmpty);
      }
    });

    test('every rule pins regulatoryRefs', () {
      for (final r in MbcDispatchCatalog.rules) {
        expect(r.regulatoryRefs, isNotEmpty,
            reason: '${r.scaleId} missing regulatory anchor');
      }
    });

    test('maxItemsPerSession > 0', () {
      for (final r in MbcDispatchCatalog.rules) {
        expect(r.maxItemsPerSession > 0, isTrue);
      }
    });

    test('schemaVersion positive', () {
      expect(MbcDispatchCatalog.schemaVersion > 0, isTrue);
    });

    test('lastReviewed is YYYY-MM', () {
      expect(MbcDispatchCatalog.lastReviewed,
          matches(RegExp(r'^\d{4}-\d{2}$')));
    });

    test('byScaleId returns the matching rule', () {
      final r = MbcDispatchCatalog.byScaleId('phq9');
      expect(r.scaleId, 'phq9');
    });

    test('byScaleId throws for unknown id', () {
      expect(() => MbcDispatchCatalog.byScaleId('nonsense'),
          throwsStateError);
    });
  });

  group('isDueForDispatch', () {
    final rule = MbcDispatchCatalog.byScaleId('phq9'); // 14-day interval

    test('returns true when never dispatched', () {
      expect(
        isDueForDispatch(
          rule: rule,
          lastDispatchedAt: null,
          now: DateTime.utc(2026, 6, 30),
        ),
        isTrue,
      );
    });

    test('returns false when still inside the interval', () {
      expect(
        isDueForDispatch(
          rule: rule,
          lastDispatchedAt: DateTime.utc(2026, 6, 25),
          now: DateTime.utc(2026, 6, 30),
        ),
        isFalse,
      );
    });

    test('returns true on the boundary', () {
      expect(
        isDueForDispatch(
          rule: rule,
          lastDispatchedAt: DateTime.utc(2026, 6, 16),
          now: DateTime.utc(2026, 6, 30),
        ),
        isTrue,
      );
    });
  });

  group('tokenExpiryFor', () {
    test('adds catalog hours to dispatchedAt', () {
      final rule = MbcDispatchCatalog.byScaleId('phq9'); // 72h
      final exp = tokenExpiryFor(
        rule: rule,
        dispatchedAt: DateTime.utc(2026, 6, 30, 12),
      );
      expect(exp, DateTime.utc(2026, 7, 3, 12));
    });
  });
}

File: functions/src/__tests__/mbc_dispatch_catalog.test.ts
import {
  MBC_DISPATCH_RULES,
  MBC_LAST_REVIEWED,
  MBC_SCHEMA_VERSION,
  isDueForDispatch,
  mbcRuleByScaleId,
  tokenExpiryMillis,
} from "../lib/mbc_dispatch_catalog";

describe("MBC dispatch catalog invariants", () => {
  test("catalog is non-empty", () => {
    expect(MBC_DISPATCH_RULES.length).toBeGreaterThan(0);
  });

  test("every rule has reminderAtHours <= linkLifetimeHours", () => {
    for (const r of MBC_DISPATCH_RULES) {
      expect(r.reminderAtHours).toBeLessThanOrEqual(r.linkLifetimeHours);
    }
  });

  test("every rule has at least one audience + channel", () => {
    for (const r of MBC_DISPATCH_RULES) {
      expect(r.audiences.length).toBeGreaterThan(0);
      expect(r.channels.length).toBeGreaterThan(0);
    }
  });

  test("every rule has non-empty regulatory anchor", () => {
    for (const r of MBC_DISPATCH_RULES) {
      expect(r.regulatoryRefs.length).toBeGreaterThan(0);
    }
  });

  test("schemaVersion + lastReviewed shape", () => {
    expect(MBC_SCHEMA_VERSION).toBeGreaterThan(0);
    expect(MBC_LAST_REVIEWED).toMatch(/^\d{4}-\d{2}$/);
  });

  test("maxItemsPerSession positive", () => {
    for (const r of MBC_DISPATCH_RULES) {
      expect(r.maxItemsPerSession).toBeGreaterThan(0);
    }
  });

  test("scaleId set is unique", () => {
    const seen = new Set<string>();
    for (const r of MBC_DISPATCH_RULES) {
      expect(seen.has(r.scaleId)).toBe(false);
      seen.add(r.scaleId);
    }
  });
});

describe("mbcRuleByScaleId", () => {
  test("returns matching rule", () => {
    expect(mbcRuleByScaleId("phq9").scaleId).toBe("phq9");
  });

  test("throws on unknown id", () => {
    expect(() => mbcRuleByScaleId("nonsense")).toThrow();
  });
});

describe("isDueForDispatch", () => {
  const rule = mbcRuleByScaleId("phq9");
  // phq9 has intervalDays = 14
  const day = 24 * 3_600_000;

  test("first-time dispatch is always due", () => {
    expect(
      isDueForDispatch({
        rule,
        lastDispatchedAtMillis: null,
        nowMillis: 1_000_000_000,
      }),
    ).toBe(true);
  });

  test("inside interval is not due", () => {
    const last = 1_000_000_000;
    const now = last + 5 * day;
    expect(
      isDueForDispatch({
        rule,
        lastDispatchedAtMillis: last,
        nowMillis: now,
      }),
    ).toBe(false);
  });

  test("at interval boundary is due", () => {
    const last = 1_000_000_000;
    const now = last + 14 * day;
    expect(
      isDueForDispatch({
        rule,
        lastDispatchedAtMillis: last,
        nowMillis: now,
      }),
    ).toBe(true);
  });
});

describe("tokenExpiryMillis", () => {
  test("phq9 link expires linkLifetimeHours after dispatch", () => {
    const rule = mbcRuleByScaleId("phq9");
    const dispatched = 1_700_000_000_000;
    const expected = dispatched + rule.linkLifetimeHours * 3_600_000;
    expect(
      tokenExpiryMillis({rule, dispatchedAtMillis: dispatched}),
    ).toBe(expected);
  });
});
